pageserver/src/page_service.rs

```diff
@@ -2086,20 +2123,81 @@ impl PageServerHandler {
                 // Ignore error (trace buffer may be full or tracer may have disconnected).
                 _ = page_trace.try_send(PageTraceEvent {
                     key,
-                    effective_lsn,
+                    effective_lsn: batch.effective_request_lsn,
                     time,
                 });
             }
         }
 
+        // If any request in the batch needs to wait for LSN, then do so now.
+        let mut perf_instrument = false;
+        let max_effective_lsn = requests
+            .iter()
+            .map(|req| {
+                if req.ctx.has_perf_span() {
+                    perf_instrument = true;
+                }
+
+                req.effective_request_lsn
+            })
```

Comment:
Bit of a smell to have side-effectful `map` functions

Reply:
Agreed, but it's avoiding one extra iteration and the `perf_instrument` declaration is right on top. Will keep as is unless you object strongly.